test/core/bad_connection/close_fd_test.cc

```diff
@@ -0,0 +1,731 @@
+  /* Create client, server and setup transport over endpoint pair */
+  init_server(ctx);
+  init_client(ctx);
+}
+
+static void drain_cq(grpc_completion_queue* cq) {
+  grpc_event ev;
+  do {
+    ev = grpc_completion_queue_next(cq, grpc_timeout_seconds_to_deadline(1),
+                                    nullptr);
+  } while (ev.type != GRPC_QUEUE_SHUTDOWN);
+}
+
+static void drain_and_destroy_cq(grpc_completion_queue* cq) {
+  grpc_completion_queue_shutdown(cq);
+  drain_cq(cq);
+  grpc_completion_queue_destroy(cq);
+}
+
+static void shutdown_server(test_ctx* ctx) {
+  if (!ctx->server) return;
+  grpc_server_shutdown_and_notify(ctx->server, ctx->shutdown_cq, tag(1000));
+  GPR_ASSERT(grpc_completion_queue_pluck(ctx->shutdown_cq, tag(1000),
+                                         grpc_timeout_seconds_to_deadline(1),
+                                         nullptr)
+                 .type == GRPC_OP_COMPLETE);
+  grpc_server_destroy(ctx->server);
+  ctx->server = nullptr;
+}
+
+static void shutdown_client(test_ctx* ctx) {
+  if (!ctx->client) return;
+  grpc_channel_destroy(ctx->client);
+  ctx->client = nullptr;
+}
+
+static void end_test(test_ctx* ctx) {
+  shutdown_server(ctx);
+  shutdown_client(ctx);
+
+  drain_and_destroy_cq(ctx->cq);
+  drain_and_destroy_cq(ctx->client_cq);
+  drain_and_destroy_cq(ctx->bound_cq);
+  grpc_completion_queue_destroy(ctx->shutdown_cq);
+  gpr_free(ctx->ep);
+}
+
+typedef enum fd_type { CLIENT_FD, SERVER_FD } fd_type;
+
+static const char* fd_type_str(fd_type f) {
+  if (f == CLIENT_FD) {
+    return "client";
+  } else if (f == SERVER_FD) {
+    return "server";
+  } else {
+    gpr_log(GPR_ERROR, "Unexpected fd_type %d", f);
+    abort();
+  }
+}
+
+static void _test_close_before_server_recv(fd_type f) {
+  grpc_core::ExecCtx exec_ctx;
+  grpc_call* c;
+  grpc_call* s;
+  grpc_event e;
+  grpc_slice request_payload_slice =
+      grpc_slice_from_copied_string("hello world");
+  grpc_slice response_payload_slice =
+      grpc_slice_from_copied_string("hello you");
+  grpc_byte_buffer* request_payload =
+      grpc_raw_byte_buffer_create(&request_payload_slice, 1);
+  grpc_byte_buffer* response_payload =
+      grpc_raw_byte_buffer_create(&response_payload_slice, 1);
+  gpr_log(GPR_INFO, "Running test: test_close_%s_before_server_recv",
+          fd_type_str(f));
+  test_ctx ctx;
+  test_init(&ctx);
+
+  grpc_op ops[6];
+  grpc_op* op;
+  grpc_metadata_array initial_metadata_recv;
+  grpc_metadata_array trailing_metadata_recv;
+  grpc_metadata_array request_metadata_recv;
+  grpc_byte_buffer* request_payload_recv = nullptr;
+  grpc_byte_buffer* response_payload_recv = nullptr;
+  grpc_call_details call_details;
+  grpc_status_code status = GRPC_STATUS__DO_NOT_USE;
+  grpc_call_error error;
+  grpc_slice details;
+
+  gpr_timespec deadline = grpc_timeout_seconds_to_deadline(1);
+  grpc_endpoint_add_to_pollset(ctx.ep->client, grpc_cq_pollset(ctx.client_cq));
```

Comment:
This looks like something that belongs to `test_init`

Reply:
Thanks, fixed.